### PR TITLE
8292879: com/sun/jdi/ClassUnloadEventTest.java failed due to classes not unloading

### PR DESCRIPTION
**Pull request by plummercj**
The bug is due to the debug agent still having a reference to one of the classes when the full GC is issued by the debuggee, thus keeping the classes loaded. The debug agent it about to free the reference, but doesn't do it quite soon enough, and then it is blocked from freeing it by the full GC. It does free it after the full GC, but by then it is too late, and the full GC has already failed to unload all the classes.

The debug agent class reference is coming from the ClassPrepareEvent that is sent for the last class loaded. The event has been sent to the debugger and the debug agent is about to free the reference, but the debuggee main thread has already resumed and gets to the full GC first. Adding a short delay in the debuggee main thread allows the debug agent to finish freeing the reference before the full GC is done.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292879](https://bugs.openjdk.org/browse/JDK-8292879): com/sun/jdi/ClassUnloadEventTest.java failed due to classes not unloading

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10519/head:pull/10519` \
`$ git checkout pull/10519`

Update a local copy of the PR: \
`$ git checkout pull/10519` \
`$ git pull https://git.openjdk.org/jdk pull/10519/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10519`

View PR using the GUI difftool: \
`$ git pr show -t 10519`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10519.diff">https://git.openjdk.org/jdk/pull/10519.diff</a>

</details>

### CODE REVIEW

File: test/jdk/com/sun/jdi/ClassUnloadEventTest.java [hunk 1]
```diff
@@ -105,10 +105,19 @@ private static void runDebuggee() {
...
+            Thread.sleep(5000);
+        } catch (InterruptedException e) {
+        }
+
         // Trigger class unloading
         ClassUnloadCommon.triggerUnloading();
 
-        // Short delay to make sure all ClassUnloadEvents have been sent
+        // Do a short delay to make sure all ClassUnloadEvents have been sent
         // before VMDeathEvent is genareated.
```
Comment (dholmes-ora) [inline test/jdk/com/sun/jdi/ClassUnloadEventTest.java:121]: Existing typo: genareated

**Review by dholmes-ora — APPROVED** [review]
This is crude but probably sufficient - and I don't know if there is any way to know for sure that it is okay to proceed.

Thanks.

**Comment by plummercj** [comment]
> This is crude but probably sufficient - and I don't know if there is any way to know for sure that it is okay to proceed.

Yeah, it is. I decided to take a better look as to whether or not this could be handled in the debug agent. The first question I found myself asking is why is there a ClassPrepareEvent being generated? The test itself does not enable them. It turns out the JDI implementation enables them, and does so with SUSPEND_NONE.

This SUSPEND_NONE suspension policy is part of the problem. It means that when the event is sent, the thread it was triggered on (the debuggee main thread) is not suspended. The event actually gets sent to the debugger from another thread (the "event helper thread"), and the class reference is freed after it is sent. But the main debugger thread considers itself "done" with the event once it gets enqueued for the event helper thread to handle, so the debugger main thread continues execution, and moves on to the full GC part of the test. Meanwhile the event helper thread is processing the event, and may take a while before it gets far enough along to free the class reference.

If the suspension policy was SUSPEND_EVENT_THREAD or SUSPEND_ALL, then the debugger main thread would have suspended until after the debugger explicitly resumed its execution. This would have allowed the sending of the event and the freeing of the class reference to complete before the debugger main thread was resumed.

I suppose another way to fix test issue would be to add an additional ClassPrepareRequest to the test, and have it use SUSPEND_EVENT_THREAD or SUSPEND_ALL. But in the end it's still just a workaround to the problem.

**Review by coleenp — APPROVED** [review]
Thanks for fixing the genareated typo.

**Review by sspitsyn — APPROVED** [review]
This is interesting.
Looks good.
Thanks,
Serguei

**Pull request closed without merging**